### PR TITLE
core: Add compatibility with PG 12 major version

### PR DESCRIPTION
**Pull request by paulrbr-fl**
This PR adds compatibility for postgresql 12 major version.

The PG12 configuration is as close as the PG11 one. All defaults have
been kept back to avoid breaking backward compatibility within this
ansible role.

Here is a complete side-by-side diff between the current PG11 conf and
the new PG12 conf:

```diff

                                        # 0 selects the syste						# 0 selects the syste
                                                              >	#tcp_user_timeout = 0			# TCP_USER_TIMEOUT, i
                                                              >						# 0 selects the syste

                                                              >	#ssl_min_protocol_version = 'TLSv1'
                                                              >	#ssl_max_protocol_version = ''

                                                              >	#shared_memory_type = mmap		# the default is the
                                                              >						# supported by the op
                                                              >						#   mmap
                                                              >						#   sysv
                                                              >						#   windows
                                                              >						# (change requires re
dynamic_shared_memory_type = posix	# the default is the    dynamic_shared_memory_type = posix	# the default is the
                                        # supported by the op						# supported by the op
                                        #   posix							#   posix
                                        #   sysv							#   sysv
                                        #   windows							#   windows
                                        #   mmap							#   mmap
                                        # use none to disable <
                                        # (change requires re						# (change requires re

wal_level = logical			# minimal, replica, o |	wal_level = {{ postgres_wal_level }}			# min
                                        # (change requires re						# (change requires re
wal_log_hints = on			# also do full page w	wal_log_hints = on			# also do full page w
                                        # (change requires re						# (change requires re
                                                              >	#wal_init_zero = on			# zero-fill new WAL f
                                                              >	#wal_recycle = on			# recycle WAL files
                                        # (change requires re						# (change requires re

                                                              >	# - Archive Recovery -
                                                              >
                                                              >	# These are only used in recovery mode.
                                                              >
                                                              >	{% if postgres_primary %}
                                                              >	{# In PG < 12 versions all the recovery settings were in a se
                                                              >	restore_command = '/usr/bin/barman-wal-restore --user barman
                                                              >					# placeholders: %p = path of
                                                              >					#               %f = file nam
                                                              >					# e.g. 'cp /mnt/server/archiv
                                                              >					# (change requires restart)
                                                              >	{% else %}
                                                              >	#restore_command = ''		# command to use to restore a
                                                              >					# placeholders: %p = path of
                                                              >					#               %f = file nam
                                                              >					# e.g. 'cp /mnt/server/archiv
                                                              >					# (change requires restart)
                                                              >	{% endif %}
                                                              >	#archive_cleanup_command = ''	# command to execute at every
                                                              >	#recovery_end_command = ''	# command to execute at compl
                                                              >
                                                              >	# - Recovery Target -
                                                              >
                                                              >	# Set these only when performing a targeted recovery.
                                                              >
                                                              >	#recovery_target = ''		# 'immediate' to end recovery
                                                              >                                 # consistent state is reached
                                                              >					# (change requires restart)
                                                              >	#recovery_target_name = ''	# the named restore point to
                                                              >					# (change requires restart)
                                                              >	#recovery_target_time = ''	# the time stamp up to which
                                                              >					# (change requires restart)
                                                              >	#recovery_target_xid = ''	# the transaction ID up to wh
                                                              >					# (change requires restart)
                                                              >	#recovery_target_lsn = ''	# the WAL LSN up to which rec
                                                              >					# (change requires restart)
                                                              >	#recovery_target_inclusive = on # Specifies whether to stop:
                                                              >					# just after the specified re
                                                              >					# just before the recovery ta
                                                              >					# (change requires restart)
                                                              >	{% if postgres_primary %}
                                                              >	{# In PG < 12 versions all the recovery settings were in a se
                                                              >	recovery_target_timeline='latest'	# 'current', 'latest'
                                                              >					# (change requires restart)
                                                              >	{% else %}
                                                              >	#recovery_target_timeline = 'latest'	# 'current', 'latest'
                                                              >					# (change requires restart)
                                                              >	{% endif %}
                                                              >	#recovery_target_action = 'pause'	# 'pause', 'promote',
                                                              >					# (change requires restart)

max_wal_senders = 10		# max number of walsender pro |	#max_wal_senders = 10		# max number of walsender pro
                                # (change requires restart)					# (change requires restart)

                                                              >	{% if postgres_primary %}
                                                              >	{# In PG < 12 versions this setting was defined in separate r
                                                              >	primary_conninfo = 'host={{ postgres_primary.host }} port={{
                                                              >						# (change requires re
                                                              >	{% else %}
                                                              >	#primary_conninfo = ''			# connection string t
                                                              >						# (change requires re
                                                              >	{% endif %}
                                                              >	#primary_slot_name = ''			# replication slot on
                                                              >						# (change requires re
                                                              >	{% if postgres_primary %}
                                                              >	{# In PG < 12 versions this setting was defined in separate r
                                                              >	promote_trigger_file = '/var/lib/postgresql/{{ postgres_versi
                                                              >	{% else %}
                                                              >	#promote_trigger_file = ''		# file name whose pre
                                                              >	{% endif %}

                                        # retrieve WAL after                                            # retrieve WAL after
                                                              >	#recovery_min_apply_delay = 0		# minimum delay for a

                                                              >	#plan_cache_mode = auto			# auto, force_generic
                                                              >						# force_custom_plan
                                                              >

                                        #   debug5	      <
                                        #   debug4	      <
                                        #   debug3	      <
                                        #   debug2	      <
                                        #   debug1	      <
                                        #   log		      <
                                        #   notice	      <
                                        #   warning	      <
                                        #   error	      <
                                                              <
log_min_duration_statement = 10000	# -1 is disabled, 0 l	log_min_duration_statement = 10000	# -1 is disabled, 0 l
                                        # and their durations						# and their durations
                                        # statements running                                            # statements running
                                        # of milliseconds						# of milliseconds

                                                              >	#log_transaction_sample_rate = 0.0	# Fraction of transac
                                                              >						# are logged regardle
                                                              >						# statements from all

                                                              >	#client_min_messages = notice		# values in order of
                                                              >						#   debug5
                                                              >						#   debug4
                                                              >						#   debug3
                                                              >						#   debug2
                                                              >						#   debug1
                                                              >						#   log
                                                              >						#   notice
                                                              >						#   warning
                                                              >						#   error

                                        # only default tables						# only default tables
                                                              >	#default_table_access_method = 'heap'

                                                              >						# selects precise out

                                                              >	#data_sync_retry = off			# retry or panic on f
                                                              >						# data?
                                                              >						# (change requires re

                                                              >	# assignments, so they can usefully be given more than once.
```

----

#

### CODE REVIEW

**Comment by paulrbr-fl** [comment]
@gaetanfl @chtitux would you folks mind reviewing this please?

File: tasks/postgres-cluster.yml [hunk 1]
```diff
@@ -82,9 +83,18 @@
...
   no_log: True
 
+- name: Define secondary nodes as secondaries with corresponding signal files
+  when: postgres_primary and postgres_version | version('12', '>=')
+  file:
+    path: {{ item }}
+    state: touch
+  with_items:
+    - /var/lib/postgresql/{{ postgres_version }}/{{ postgres_cluster_name }}/standby.signal
+    - /var/lib/postgresql/{{ postgres_version }}/{{ postgres_cluster_name }}/recovery.signal
```
Comment (chtitux) [inline tasks/postgres-cluster.yml]: I think we should keep only standby.signal, as we don't want the recovery mode.

Comment (paulrbr-fl) [reply tasks/postgres-cluster.yml]: indeed, thanks I removed it :+1:

**Comment by paulrbr-fl** [comment]
@gaetanfl & @chtitux I have enabled the automatic tests on Travis for PG 10, 11 & 12. This validates the basic usage of the role (and the new PG 12 configuration added in this PR).

I didn't add tests for PG 9.5 & 9.6 because of errors I didn't take the time to solve but that I detailed in [this commit](https://github.com/trainline-eu/ansible-postgresql-role/pull/11/commits/61b4552ae3514a7ae94591a85c4708ac0d7db078).
I think we can move forward and merge this PR and let another PR fix the tests for 9.5 & 9.6.

WDYT?

**Review by gaetanfl — APPROVED** [review]
LGTM

**Comment by paulrbr-fl** [comment]
Thanks both for the review!

**Comment by paulrbr-fl** [comment]
Thanks both for the review!

**Pull request merged**